libparsec/crates/client/src/workspace/store/reparent_updater.rs

```diff
@@ -0,0 +1,203 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::sync::Arc;
+
+use libparsec_types::prelude::*;
+
+use crate::workspace::store::resolve_path::ResolvePathForReparentingError;
+use libparsec_platform_storage::workspace::UpdateManifestData;
+
+use super::per_manifest_update_lock::ManifestUpdateLockGuard;
+
+pub(super) type UpdateManifestsForReparentingError = super::WorkspaceStoreOperationError;
+pub(crate) type ForUpdateReparentingError = ResolvePathForReparentingError;
+
+pub(super) async fn resolve_path_for_update_reparenting<'a>(
+    store: &'a super::WorkspaceStore,
+    src_parent_path: &FsPath,
+    src_child_name: &EntryName,
+    dst_parent_path: &FsPath,
+    // In most reparenting we don't even care if the destination exists given
+    // it is simply overwritten by the source.
+    // However this is not the case when doing an exchange between source and
+    // destination, given in this case the destination's `parent` field must
+    // also be modified.
+    dst_child_name: Option<&EntryName>,
+) -> Result<ReparentingUpdater<'a>, ForUpdateReparentingError> {
+    let resolution = super::resolve_path::resolve_path_for_reparenting(
+        store,
+        src_parent_path,
+        src_child_name,
+        dst_parent_path,
+        dst_child_name,
+    )
+    .await?;
+
+    Ok(ReparentingUpdater {
+        src_parent_manifest: resolution.src_parent_manifest,
+        src_child_manifest: resolution.src_child_manifest,
+        dst_parent_manifest: resolution.dst_parent_manifest,
+        dst_child_manifest: resolution.dst_child_manifest,
+        guards: ReparentingUpdaterGuards {
+            store,
+            update_guards: Some((
+                resolution.src_parent_update_lock_guard,
+                resolution.src_child_update_lock_guard,
+                resolution.dst_parent_update_lock_guard,
+                resolution.dst_child_update_lock_guard,
+            )),
+        },
+    })
+}
+
+pub(crate) struct ReparentingUpdater<'a> {
+    pub src_parent_manifest: Arc<LocalFolderManifest>,
+    pub src_child_manifest: ArcLocalChildManifest,
+    pub dst_parent_manifest: Arc<LocalFolderManifest>,
+    pub dst_child_manifest: Option<ArcLocalChildManifest>,
+    guards: ReparentingUpdaterGuards<'a>,
+}
+
+/// Destructuring and `impl Drop` are not compatible, hence this structure that is
+/// responsible to drop the update guards, while `ReparentingUpdater` can still be
+/// destructured (which is done in `update_manifests` to populate the cache with
+/// it manifests).
+struct ReparentingUpdaterGuards<'a> {
+    store: &'a super::WorkspaceStore,
+    update_guards: Option<(
+        ManifestUpdateLockGuard,
+        ManifestUpdateLockGuard,
+        ManifestUpdateLockGuard,
+        Option<ManifestUpdateLockGuard>,
```

Comment:
Are these guards the guards of each manifest in ReparentingUpdater ?

Reply:
yes, the `ManifestUpdateLockGuard` are low-level guard that are not released on drop

Given they must be handled with extra care, we have the Updater structures that wrap them and allow better ergonomics